### ISSUE TITLE
Remove old logs based on a maximum number of files

### ISSUE THREAD

**Issue by DRoet** [feugy/pino-roll#8]
Hi @feugy 

I don't believe removing old logs is currently supported, right? I think you mentioned it in an earlier [comment](https://github.com/pinojs/pino/issues/1323#issuecomment-1084299639).
Not really sure what is the best way to implement this but I figured I would create this issue to keep it on the radar.

**Comment by burgerni10** [comment]
Hi @feugy, do you need help implementing that ? 

Since you already mentioned it, do you have some hints on how to implement it properly ?

**Comment by feugy** [comment]
Hello folks, apologies for the long silence.
This is an handy feature indeed. @burgerni10 I'll give it a try today, but in case you're faster than me, a _possible_ plan would be to:
1. maintain (in memory) the current number of log files.
2. on log rotation (`scheduleRoll()` function), asynchronously trigger a file deletion, if needed (we don't want to delay the rotation with deleting a file)
3. on startup, when this feature is enabled, count the number of log file (possibly asynchronously as weel, it sounds a acceptable compromise)

**Comment by feugy** [comment]
Hello gents.
Let me amend my thinking here.

_Removing old log files_ means crawling files in the same folder, count them, order them by something (file name, or access date), then remove some. It's raising some questions around identifying older log files (matching the same file name pattern? all of them?)
It could lead to some undesirable, catastrophic behavior: configure a limit of 1, start you app in a folder full of old log files, and they are all gone.

_Limiting the number of created files_ is simpler and safer to reason about: during the same execution, pino-roll limits the number of created file, and simply ignores what was in the folder before.

In my opinion, this allows the same level of control, without the risk of removing other files.
This is implemented in #38: `limit.count` is a new parameter (we can imagine limit by age or size as well)

Is that acceptable?

**Comment by burgerni10** [comment]
Hi @feugy, thank you for the investigation.

Your solution has the drawback of having unlimited files in case of application restart: if limit.count is set to 10, and the app restart at count equals to 9, it will keep indefinitely these files on disk. Other restarts in the future will increase the number of files on the disk indefinitely.



I suggest reading the folder that match the following regex: `file[0-9].*\.log` where log is the extension, and file is the filename.

Let's suppose 3 files are in the folder:
- file1.log (size 20 MB, modified 20min ago)
- file2.log (size 10 MB, modified 2min ago)
- file3.log (size 20 MB, modified 40min ago)

And pino-roll set up with 
```json
"options": {
  "file": "file",
  "size": 20,
  "limit": { "count": 3}
}
```

The most recent file is file2.log and its size is inferior to the size option. Thus, it will be reopen by the logger at startup. 

When file2.log is full, file3.log will be removed and open a new. I suggest to overwrite an old file at new file creation only.
When the new file3.log will be full, since limit.count is set to 3, next file to be overwritten will be file1.log

This way, we limit the number of file on disk by design and we do not remove old logs at startup.
The only way to keep a file on disk indefinitely is to reduce the limit.count from 3 to 2 for example when a file3.log has already been created. This seems more acceptable to me.

What do you think ?

**Comment by burgerni10** [comment]
> the risk of messing up with existing files (which is mitigated with applying file name filtering as you suggested)

I agree there is a risk, but I think the risk is acceptable since it will depend on a filename that must be set by the user.

> my "vision" of pino-roll, compared to existing system tools like logrotate. Fundamentally, pino-roll is an handy tool for development, prototyping and experimenting services. If one need something bullet-proof and production-grade, stdout + logrotate is the way to go.

I didn't have the vision at first, but I get it now :) 
The use of logrotate for production-grade totally makes sense. However, the logger may be embedded in a cross-platform NodeJS app (that is my use case).  In such cases, using logrotate implies to depend of an external tool (especially for Windows). 

I think having a logrotate logic embedded in pino-roll makes sense in order to have an all-included feature, as it is done in [winston](https://github.com/winstonjs/winston-daily-rotate-file). Maybe I am mistaking, but I think pino-roll is not far from having such a feature (which I can help to develop of course), and it would be a shame to develop a similar transport for that.

**Comment by burgerni10** [comment]
Hi @feugy , any thoughts on my last comments ?

**Comment by MGough** [comment]
Hello! I'm in agreement with @burgerni10 - I can see the concerns that you have @feugy, but I think that there's definitely a separate use-case here, and that perhaps it'd be reasonable to support both deleting pre-existing files, or starting fresh on new runs. I'm also working with an embedded application which is running on the user's own machine, pino-roll is almost perfect for it, particularly as I cannot use other logging libraries due to separate constraints. I really need to ensure I'm not filling up their filesystem over time, while also enabling them to send the most recent logfiles to me in case of application error.

**Comment by bkvishe** [comment]
Can we make this configurable? Like one can set a flag on/off to control the removal of existing files on application restart.

**Comment by mcollina** [comment]
@bkvishe sure, would you like to send a PR?

**Comment by bkvishe** [comment]
Sure. I Will give it a try.

**Comment by benedictjohannes** [comment]
It's been some months and I'm interested in the same feature. I'd use `limit.removeOtherFiles`, which accepts the values `'name' | 'modified' | true`. Setting `name` will remove the first files with the ordered using natural order ascending, while setting `modified` will remove the files with the oldest modified files. `true` will default to modified, I guess.

I'll be sure to add a warning that usage of this option should only be specified when the folder is only used for logging and older log files are always safe to be removed.

**Comment by benedictjohannes** [comment]
Hi @mcollina I'm starting work on my proposal to add the option `limit.removeOtherFiles` with valid values `'name'|'modified'|true`

Hopefully I can open another PR soon. I'd hope you agree with this design, as this is opt-in, there should be no concern about deleting other files that's not created by the current instance of logger.

**Comment by mcollina** [comment]
make sense
